lib/plausible_web/controllers/auth_controller.ex

```diff
@@ -550,20 +552,42 @@ defmodule PlausibleWeb.AuthController do
 
   def update_email(conn, %{"user" => user_params}) do
     user = conn.assigns.current_user
-    changes = Auth.User.email_changeset(user, user_params)
 
-    case Repo.update(changes) do
-      {:ok, user} ->
-        if user.email_verified do
-          handle_email_updated(conn)
-        else
-          Auth.EmailVerification.issue_code(user)
-          redirect(conn, to: Routes.auth_path(conn, :activate_form))
+    case RateLimit.check_rate(
+           "email-change:user:#{user.id}",
+           @email_change_interval,
+           @email_change_limit
+         ) do
+      {:allow, _} ->
+        changes = Auth.User.email_changeset(user, user_params)
+
+        case Repo.update(changes) do
+          {:ok, user} ->
+            if user.email_verified do
+              handle_email_updated(conn)
+            else
+              Auth.EmailVerification.issue_code(user)
+              redirect(conn, to: Routes.auth_path(conn, :activate_form))
+            end
+
+          {:error, changeset} ->
+            settings_changeset = Auth.User.settings_changeset(user)
+
+            render_settings(conn,
+              settings_changeset: settings_changeset,
+              email_changeset: changeset
+            )
         end
 
-      {:error, changeset} ->
+      {:deny, _} ->
         settings_changeset = Auth.User.settings_changeset(user)
 
+        {:error, changeset} =
+          user
+          |> Auth.User.email_changeset(user_params)
```

Comment:
Maybe change it to:

```suggestion
          |> Auth.User.settings_changeset()
```

to reset the inputs, as in the case of standard changeset error?

Reply:
`settings_changeset` is reset at line 583, I guess that's what you meant?